source/agora/common/TransactionManager.d

```diff
@@ -0,0 +1,211 @@
+/*******************************************************************************
+
+    Contains class for transaction verification
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.TransactionManager;
+import agora.common.Data;
+import agora.common.crypto.Key;
+import agora.common.Transaction;
+import agora.common.Hash;
+import  std.algorithm;
+
+/*******************************************************************************
+
+Class for the transactions verification
+
+*******************************************************************************/
+
+public class TransactionManager
+{
+    /// Store of previously processed transactions
+    private Transaction[Hash] txs;
+
+    /*******************************************************************************
+
+        Save a `Transaction`
+
+        Params:
+            hash = `Hash` of `Transaction`
+            tx = Type of `Transaction`
+
+    *******************************************************************************/
+
+    public void saveTransaction (Hash hash, Transaction tx)
+    {
+        this.txs[hash] = tx;
+    }
+
+    /*******************************************************************************
+
+        Return hash of transaction
+
+        Params:
+            tx = Type of `Transaction`
+
+        Return:
+            `Hash` of `Transaction`
+
+    *******************************************************************************/
+
+    public Hash getTxHash (Transaction tx)
```

Comment:
Do you plan on ever overriding and changing what this method does? Because if not, we don't really need it. Unittest code can simply call `hashFull(tx)`.

Reply:
> Great work !
> 
> One point though, if a function does not use any member of a class, and its usage is not specific to a class, it should not be part of a class. Other parts of the program might need to get the `sum` of the input. They should not have to have an instance of a `TransactionManager` to do so.
> 
> For example, imagine I want to print the sum of all transactions in a debug logging statement in the network code.
> 
> Using
> 
> https://github.com/bpfkorea/agora/blob/740d54a677a209dabcb8fbd1893c8f2950089b37/source/agora/node/Network.d#L300-L306
> 
> as a base
> ```d
>     public void sendMessage (Transaction tx) @safe
>     {
>         scope tm = new TransactionManager();
>         foreach (ref node; this.peers)
>         {
>             logDebug("Sending a transaction with %d input and %d output", tm.getInputSum(tx), tm.getOutputSum(tx));
>             node.sendMessage(tx);
>         }
>     }
> ```
> 
> Instead, if you make it a free function:
> 
> ```d
> public long getOutputSum (Transaction tx)
> {
>     return tx.outputs.map!(a => a.value).sum();
> }
> ```
> 
> I can call it like this:
> 
> ```d
>     public void sendMessage (Transaction tx) @safe
>     {
>         foreach (ref node; this.peers)
>         {
>             logDebug("Sending a transaction with %d input and %d output", tx.getInputSum(), tx.getOutputSum());
>             node.sendMessage(tx);
>         }
>     }
> ```
> 
> Which I think is much clearer. It does not have to be a member of `Transaction` because we can use UFCS, as you already did with `map` and `sum`.

The function is dependent on the internal data structure.
Please check in detail to see if you can do this.
Wouldn't it be better to make improvements after the blocks are made in the future?

Reply:
Modified~